File: backend/src/main/java/woorifisa/project/backend/domain/job/entity/Job.java
package woorifisa.project.backend.domain.job.entity;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import woorifisa.project.backend.global.entity.BaseEntity;

@Getter
@Entity
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Table(name = "job")
public class Job extends BaseEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "job_id")
    private Long jobId;

    @Column(name = "company", length = 100, nullable = false)
    private String company;

    @Column(name = "region", length = 100, nullable = false)
    private String region;

    @Column(name = "opening_title", length = 100, nullable = false)
    private String openingTitle;

    @Column(name = "job_category", length = 50)
    private String jobCategory;

    @Column(name = "experience", length = 50)
    private String experience;

    @Column(name = "salary", length = 50)
    private String salary;

    @Column(name = "deadline_type", length = 50)
    private String deadlineType;

    @Column(name = "recruit_count", length = 50)
    private String recruitCount;

    @Column(name = "preferred", length = 100)
    private String preferred;

    @Column(name = "age", length = 50)
    private String age;

    @Column(name = "gender", length = 50)
    private String gender;

    @Column(name = "job_role", length = 50)
    private String jobRole;

    @Column(name = "work_period", length = 50)
    private String workPeriod;

    @Column(name = "employment_type", length = 50)
    private String employmentType;

    @Column(name = "benefits", length = 100)
    private String benefits;

    @Column(name = "address", length = 255)
    private String address;

    @Column(name = "introduce", columnDefinition = "TEXT")
    private String introduce;
}

File: frontend/src/api/endpoints/job.ts
import apiClient from '../client'

type ApiEnvelope<T> = {
  success: boolean
  code: string
  message: string
  data: T
}

export type JobOpeningItemResponse = {
  job_id: number
  company: string
  region: string
  opening_title: string
  job_category: string
  experience: string
  work_period: string
  salary: string
  created_at: string
}

export type JobOpeningListResponse = {
  items: JobOpeningItemResponse[]
  page: number
  size: number
  has_next: boolean
}

export type JobOpeningResponse = JobOpeningItemResponse & {
  deadline_type: string
  recruit_count: string
  preferred: string
  age: string
  gender: string
  job_role: string
  employment_type: string
  benefits: string
  address: string
  introduce: string
}

export type ApplicationFormPortfolioResponse = {
  portfolio_id: number
  name: string
  url: string
}

export type ApplicationFormResponse = {
  user_id: number
  name: string
  email: string
  portfolios: ApplicationFormPortfolioResponse[]
}

export type ApplicationStatus = 'PASSED' | 'FAILED' | 'UNREAD' | 'READ'

export type ApplicationItemResponse = {
  application_id: number
  job_id: number
  opening_title: string
  applied_at: string
  status: ApplicationStatus
}

export type ApplicationListResponse = {
  items: ApplicationItemResponse[]
  page: number
  size: number
  has_next: boolean
}

export type ApplicationPortfolioItemResponse = {
  name: string
  url: string
}

export type ApplicationPortfolioResponse = ApplicationPortfolioItemResponse[]

export type JobOpeningListParams = {
  page?: number
  size?: number
}

export const jobApi = {
  listOpenings: async (params: JobOpeningListParams = {}): Promise<JobOpeningListResponse> => {
    const response = await apiClient.get<ApiEnvelope<JobOpeningListResponse>>('/jobs', {
      params: {
        page: params.page ?? 0,
        size: params.size ?? 50,
      },
    })
    return response.data.data
  },

  getOpening: async (jobId: number): Promise<JobOpeningResponse> => {
    const response = await apiClient.get<ApiEnvelope<JobOpeningResponse>>(`/jobs/${jobId}`)
    return response.data.data
  },

  getApplicationForm: async (): Promise<ApplicationFormResponse> => {
    const response =
      await apiClient.get<ApiEnvelope<ApplicationFormResponse>>('/jobs/applications/form')
    return response.data.data
  },

  listApplications: async (
    params: JobOpeningListParams = {}
  ): Promise<ApplicationListResponse> => {
    const response = await apiClient.get<ApiEnvelope<ApplicationListResponse>>(
      '/jobs/applications',
      {
        params: {
          page: params.page ?? 0,
          size: params.size ?? 50,
        },
      }
    )
    return response.data.data
  },

  getApplicationPortfolio: async (
    applicationId: number
  ): Promise<ApplicationPortfolioResponse> => {
    const response = await apiClient.get<
      ApiEnvelope<ApplicationPortfolioResponse | ApplicationPortfolioItemResponse | null>
    >(
      `/jobs/applications/${applicationId}/portfolios`
    )
    const portfolios = response.data.data

    if (!portfolios) {
      return []
    }

    return Array.isArray(portfolios) ? portfolios : [portfolios]
  },

  submitApplication: async (
    jobId: number,
    payload: { portfolioUrls: string[]; files: File[] }
  ): Promise<void> => {
    const formData = new FormData()
    formData.append(
      'body',
      new Blob([JSON.stringify({ portfolio_urls: payload.portfolioUrls })], {
        type: 'application/json',
      })
    )
    payload.files.forEach((file) => formData.append('files', file))

    await apiClient.post<ApiEnvelope<null>>(`/jobs/${jobId}/applications`, formData)
  },
}
